services/archiver/archiver.go

```diff
@@ -0,0 +1,306 @@
+// Copyright 2020 The Gitea Authors.
+// All rights reserved.
+// Use of this source code is governed by a MIT-style
+// license that can be found in the LICENSE file.
+
+package archiver
+
+import (
+	"io"
+	"io/ioutil"
+	"os"
+	"path"
+	"regexp"
+	"strings"
+	"sync"
+	"time"
+
+	"code.gitea.io/gitea/modules/base"
+	"code.gitea.io/gitea/modules/context"
+	"code.gitea.io/gitea/modules/git"
+	"code.gitea.io/gitea/modules/log"
+	"code.gitea.io/gitea/modules/setting"
+
+	"github.com/unknwon/com"
+)
+
+// ArchiveRequest defines the parameters of an archive request, which notably
+// includes the specific repository being archived as well as the commit, the
+// name by which it was requested, and the kind of archive being requested.
+// This is entirely opaque to external entities, though, and mostly used as a
+// handle elsewhere.
+type ArchiveRequest struct {
+	uri             string
+	repo            *git.Repository
+	refName         string
+	ext             string
+	archivePath     string
+	archiveType     git.ArchiveType
+	archiveComplete bool
+	commit          *git.Commit
+	cchan           chan struct{}
+}
+
+var archiveInProgress []*ArchiveRequest
+var archiveMutex sync.Mutex
+
+// SHA1 hashes will only go up to 40 characters, but SHA256 hashes will go all
+// the way to 64.
+var shaRegex = regexp.MustCompile(`^[0-9a-f]{4,64}$`)
+
+// These facilitate testing, by allowing the unit tests to control (to some extent)
+// the goroutine used for processing the queue.
+var archiveQueueMutex *sync.Mutex
+var archiveQueueStartCond *sync.Cond
+var archiveQueueReleaseCond *sync.Cond
+
+// GetArchivePath returns the path from which we can serve this archive.
+func (aReq *ArchiveRequest) GetArchivePath() string {
+	return aReq.archivePath
+}
+
+// GetArchiveName returns the name of the caller, based on the ref used by the
+// caller to create this request.
+func (aReq *ArchiveRequest) GetArchiveName() string {
+	return aReq.refName + aReq.ext
+}
+
+// IsComplete returns the completion status of this request.
+func (aReq *ArchiveRequest) IsComplete() bool {
+	return aReq.archiveComplete
+}
+
+// WaitForCompletion will wait for this request to complete, with no timeout.
+// It returns whether the archive was actually completed, as the channel could
+// have also been closed due to an error.
+func (aReq *ArchiveRequest) WaitForCompletion() bool {
+	<-aReq.cchan
+	return aReq.IsComplete()
```

Comment:
```suggestion
	import "code.gitea.io/gitea/modules/graceful"

	graceful.GetManager().RunWithShutdownContext(func(ctx context.Context) {
		select {
		case <-aReq.cchan:
			return aReq.IsComplete()
		case <-ctx.Done():
			return false
		}
	}
```
We should not block Gitea from shutting down. 😄 
Note: the `import` goes at the top of the file, of course.

Reply:
I think this (and the timeout comment below) is addressed, but I'll leave it unresolved for now in case there's follow-up -- I'm not 100% certain about the approach I ended up taking, but I think it's OK since we'll still terminate and promptly close the locally-constructed wait channel.